msal_extensions/persistence.py

```diff
@@ -197,19 +201,17 @@ class KeychainPersistence(BasePersistence):
     and protected by native Keychain libraries on OSX"""
     is_encrypted = True
 
-    def __init__(self, signal_location, service_name, account_name):
+    def __init__(self, signal_location, service_name=None, account_name=None):
         """Initialization could fail due to unsatisfied dependency.
 
         :param signal_location: See :func:`persistence.LibsecretPersistence.__init__`
         """
-        if not (service_name and account_name):  # It would hang on OSX
-            raise ValueError("service_name and account_name are required")
         from .osx import Keychain, KeychainError  # pylint: disable=import-outside-toplevel
         self._file_persistence = FilePersistence(signal_location)  # Favor composition
         self._Keychain = Keychain  # pylint: disable=invalid-name
         self._KeychainError = KeychainError  # pylint: disable=invalid-name
-        self._service_name = service_name
-        self._account_name = account_name
+        self._service_name = service_name or _auto_hash(signal_location)
+        self._account_name = account_name or _auto_hash(signal_location)
```

Comment:
Do we want to use 

1. different `service_name`s for MSAL token cache and service principal entries?
2. same `service_name` and `account_name` for one `signal_location`?

Perhaps we can default `service_name` to `_auto_hash('msal-extensions-python')`? Just a sudden idea.

Reply:
> Do we want to use
>
> 1. different `service_name`s for MSAL token cache and service principal entries?

Keep in mind that, in this persistence layer, the class `KeychainPersistence` does not know - and does not need to know - whether the data would be MSAL's token cache or service principal entries.

How would you call and use this `KeychainPersistence`? If you have 2 separate instances as `msal_token_cache = KeychainPersistence("/path/to/msal_token_cache.bin")` and `sp_entries = KeychainPersistence("/path/to/az_cli_sp.bin")`, they will naturally have different `service_name` and different `account_name` after this PR got merged.

> 2. same service_name and account_name for one signal_location?

This PR is already doing that. Same `signal_location` would hash into same `service_name`. Same applies to `account_name`.

> 3. Perhaps we can default `service_name` to `_auto_hash('msal-extensions-python')`?

If your point is to use a fixed `service_name`, we could do that. In that case we won't have to hash. We could just use default service_name as `msal-extensions` (note that I removed the `-python` suffix in your idea, because MSAL EX in other language may still want to potentially share the keychain).

I really do not know which one is better, though: using one fixed default value, or using a dynamical hash based on the `signal_location`.

Reply:
> I really do not know which one is better, though: using one fixed default value, or using a dynamical hash based on the `signal_location`.

This is exactly my point! I don't know if `service_name` should be fixed. Perhaps discuss with @xiangyan99 ? 🤣

Here is `azure-identity`'s code:

https://github.com/Azure/azure-sdk-for-python/blob/47108e9251cd706f48bea359a388b6c8fd161faf/sdk/identity/azure-identity/azure/identity/_persistent_cache.py#L81-L84

```py
    if sys.platform.startswith("darwin"):
        # the cache uses this file's modified timestamp to decide whether to reload
        file_path = os.path.expanduser(os.path.join("~", ".IdentityService", cache_name))
        return msal_extensions.KeychainPersistence(file_path, "Microsoft.Developer.IdentityService", account_name)
```

which `service_name` is fixed to `Microsoft.Developer.IdentityService`.

Reply:
Well then, in the latest amendment, we also use the same `Microsoft.Developer.IdentityService` (for the lack of a better name - naming is hard).